### PR TITLE
modbus: serial: Fix incomplete transmission issue

### LINKED ISSUE #40578: MODBUS RS-485 transceiver support broken on several platforms due to DE race condition
**Describe the bug**
When using modbus on the atsamd21 sam0 platform, the modbus uart driver de-asserts the DE pin on re-entry into the cb_handler_tx after the last byte has been written to uart_fifo_fill [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/modbus/modbus_serial.c#L391).
However, this creates a race condition since after the last byte is written to the TX register and begins clocking out, the isr re-entry and modbus_serial_tx_off causes DE is de-asserted before the byte is fully transmitted.
This results in truncation of data on the last byte when using rs495 half duplex transceivers on the differential output.
The following output on low baudrates first identified in issue #38264.
131894058-111a0370-e119-4c23-9433-86a04d393831

Since the implementation of modbus expects data to be fully transmitted before modbus_serial_tx_off is called, modbus should explicitly check if tx has completed. The following code snippit along with the sam0 uart changes outlined in PR #38311 were able to successfully resolve the issue however others have reported on different platforms that this fix does not work.
```
diff --git a/subsys/modbus/modbus_serial.c b/subsys/modbus/modbus_serial.c
index 06e44bd241..9f5dddac5f 100644
--- a/subsys/modbus/modbus_serial.c
+++ b/subsys/modbus/modbus_serial.c
@@ -393,7 +393,7 @@ static void cb_handler_tx(struct modbus_context *ctx)
                                   cfg->uart_buf_ctr);
                cfg->uart_buf_ctr -= n;
                cfg->uart_buf_ptr += n;
-       } else {
+       } else if (uart_irq_tx_complete(cfg->dev) != 0) {
                /* Disable transmission */
                cfg->uart_buf_ptr = &cfg->uart_buf[0];
                modbus_serial_tx_off(ctx);
```

**To Reproduce**
Attach oscilloscope to TX and DE pins on the sam0 platform with 115200 baudrate.
Trigger a data transmission via the modbus driver.
observe on oscilloscope DE is de-asserted before the last byte is transmitted as shown in the image above.

**Expected behavior**
DE should de-assert after the last byte has finished being transmitted as shown below with the changes outlined above in this issue on sam0.

132058110-afdf0b70-995b-4af5-8e2e-9f8758693912

**Impact**
Modbus is non functional on hardware platform implementations that differentiate between uart_irq_tx_ready and uart_irq_tx_complete. This is due to modbus checking [uart_irq_tx_ready](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/modbus/modbus_serial.c#L422) before calling the cb_handler_tx funciton. cb_handler_tx is entered when the uart TX register is empty but the uart peripheral shift register has not finished shifting out the bits yet resulting in the issue described above.
Platforms which utilize the same underlying driver layer such as the nRF52 for both [ irq_tx_ready](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/serial/uart_nrfx_uarte.c#L1639) and [irq_tx_complete ](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/serial/uart_nrfx_uarte.c#L1642) methods are unaffected.

**Logs and console output**
Pictures shown above.

**Environment (please complete the following information):**
 - OS: Windows
 - Toolchain Zephyr SDK
 - Commit SHA or Version used:
   Original commit fix on my fork [9d2758b8a7](https://github.com/rocky134/zephyr/tree/9d2758b8a7/)
   From zephyr RTOS base commit d33017b458fbdb9974f70939f099be5bfcf35794

**Additional context**
The fix described in this issue was identified as broken on other platforms here https://github.com/zephyrproject-rtos/zephyr/pull/38311#discussion_r754478074 which might have similar hidden bugs in the underlying uart implementation to be found.
Similarly, when first diagnosing this issue, I also had a similar problem where the RTOS would hang infinitely in isr checking for tx_complete and it would never be set.
This was due to the [INTFLAG register being |= with a bitmask](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/serial/uart_sam0.c#L883)  in the uart_sam0_irq_update function. Per the datasheet writing a '1' to any register in INTFLG clears it therefore `|=` resolving into `INTFLAG = INTFLAG | bitmask` would clear all pending interrupts resulting in any later checks with `regs->INTFLAG.bit.TXC != 0` to always return 0. This fix for the sam0 was also included in PR #38311
This issue affecting modbus being at a higher level would need to be checked against multiple uart hardware implementations to ensure a similar bug does not exist which might go un-noticed otherwise.

### PR DESCRIPTION
**Pull request by mcscholtz**
This patch is to fix an issue where the transceiver chip is
disabled before it has transmitted all data. This causes the
message to be corrupted because the last few bytes are missing.

The fix adds a check to make sure the transmission is completed
before disabling the transceiver.

Signed-off-by: Marius Scholtz <mariuss@ricelectronics.com>

@jfischer-no I do not have the permission to re-open the [PR ](https://github.com/zephyrproject-rtos/zephyr/pull/42051)

I have re-opened it here.

### CODE REVIEW

File: subsys/modbus/modbus_serial.c [hunk 1]
```diff
@@ -393,12 +393,22 @@ static void cb_handler_tx(struct modbus_context *ctx)
 				   cfg->uart_buf_ctr);
 		cfg->uart_buf_ctr -= n;
 		cfg->uart_buf_ptr += n;
-	} else {
-		/* Disable transmission */
-		cfg->uart_buf_ptr = &cfg->uart_buf[0];
-		modbus_serial_tx_off(ctx);
-		modbus_serial_rx_on(ctx);
+		return;
```
Comment (jfischer-no) [inline subsys/modbus/modbus_serial.c:396]: Can you please change it to this? What, IMO, in this case looks a little readable.

``` diff c
diff --git a/subsys/modbus/modbus_serial.c b/subsys/modbus/modbus_serial.c
index 2c7cb15e027..4dcc602885c 100644
--- a/subsys/modbus/modbus_serial.c
+++ b/subsys/modbus/modbus_serial.c
@@ -366,7 +366,14 @@ static void cb_handler_tx(struct modbus_context *ctx)
                                   cfg->uart_buf_ctr);
                cfg->uart_buf_ctr -= n;
                cfg->uart_buf_ptr += n;
-       } else {
+               return;
+       }
+
+       /* Must wait till the transmission is complete or
+        * RS-485 tranceiver could be disabled before all data has
+        * been transmitted and message will be corrupted.
+        */
+       if (uart_irq_tx_complete(cfg->dev)) {
                /* Disable transmission */
                cfg->uart_buf_ptr = &cfg->uart_buf[0];
                modbus_serial_tx_off(ctx);
```

Comment (jfischer-no) [reply subsys/modbus/modbus_serial.c:397]: @mcscholtz It does not look right, please recheck :-)

Comment (mcscholtz) [reply subsys/modbus/modbus_serial.c:396]: My bad, it should be good now.

**Review by carlescufi — APPROVED** [review]

**Review by jfischer-no — APPROVED** [review]
@mcscholtz Thanks for the patience

**Pull request merged**